File: aspect_extraction/extractors/rule_based.py
"""Rule-based aspect extractor using linguistic patterns."""

import re
from typing import Dict, List, Optional, Set

from aspect_extraction.core.aspect import Aspect, Sentiment
from aspect_extraction.core.extractor import AspectExtractor


class RuleBasedExtractor(AspectExtractor):
    """Rule-based aspect extractor using dependency parsing and POS patterns.

    This extractor uses linguistic rules to identify aspects based on:
    - Noun phrases (potential aspects)
    - Dependency relations (subject, object, modifier)
    - POS tag patterns (e.g., ADJ + NOUN)
    - Sentiment lexicons for polarity detection

    The advantage of this approach is speed and interpretability,
    though it may have lower recall than ML-based methods.

    Args:
        use_spacy: Whether to use spaCy for advanced parsing (default: True)
        sentiment_lexicon: Optional custom sentiment lexicon

    Example:
        >>> extractor = RuleBasedExtractor()
        >>> text = "The camera quality is excellent but battery life is poor"
        >>> aspects = extractor.extract(text)
        >>> for aspect in aspects:
        ...     print(f"{aspect.text}: {aspect.sentiment}")
        camera quality: positive
        battery life: negative
    """

    def __init__(
        self,
        use_spacy: bool = True,
        sentiment_lexicon: Optional[Dict[str, Sentiment]] = None,
    ) -> None:
        """Initialize the rule-based extractor.

        Args:
            use_spacy: Whether to use spaCy for NLP processing
            sentiment_lexicon: Custom sentiment words mapping
        """
        self.use_spacy = use_spacy
        self._nlp = None

        # Default sentiment lexicon
        self._sentiment_lexicon = sentiment_lexicon or self._get_default_sentiment_lexicon()

        if use_spacy:
            try:
                import spacy

                try:
                    self._nlp = spacy.load("en_core_web_sm")
                except OSError:
                    # Model not downloaded, fall back to non-spacy mode
                    self.use_spacy = False
            except ImportError:
                # spaCy not installed
                self.use_spacy = False

    def _get_default_sentiment_lexicon(self) -> Dict[str, Sentiment]:
        """Get default sentiment lexicon.

        Returns:
            Dictionary mapping words to sentiment polarities
        """
        return {
            # Positive words
            "excellent": Sentiment.POSITIVE,
            "great": Sentiment.POSITIVE,
            "good": Sentiment.POSITIVE,
            "amazing": Sentiment.POSITIVE,
            "wonderful": Sentiment.POSITIVE,
            "fantastic": Sentiment.POSITIVE,
            "perfect": Sentiment.POSITIVE,
            "love": Sentiment.POSITIVE,
            "best": Sentiment.POSITIVE,
            "outstanding": Sentiment.POSITIVE,
            "superb": Sentiment.POSITIVE,
            "awesome": Sentiment.POSITIVE,
            "brilliant": Sentiment.POSITIVE,
            # Negative words
            "poor": Sentiment.NEGATIVE,
            "bad": Sentiment.NEGATIVE,
            "terrible": Sentiment.NEGATIVE,
            "awful": Sentiment.NEGATIVE,
            "horrible": Sentiment.NEGATIVE,
            "worst": Sentiment.NEGATIVE,
            "disappointing": Sentiment.NEGATIVE,
            "disappoints": Sentiment.NEGATIVE,
            "hate": Sentiment.NEGATIVE,
            "useless": Sentiment.NEGATIVE,
            "mediocre": Sentiment.NEGATIVE,
            "subpar": Sentiment.NEGATIVE,
        }

    def extract(self, text: str) -> List[Aspect]:
        """Extract aspects from text using rule-based patterns.

        Args:
            text: Input text to analyze

        Returns:
            List of extracted aspects with sentiment

        Raises:
            ValueError: If text is empty or invalid
        """
        if not text or not text.strip():
            raise ValueError("Input text cannot be empty")

        if self.use_spacy and self._nlp is not None:
            return self._extract_with_spacy(text)
        else:
            return self._extract_with_patterns(text)

    def _extract_with_spacy(self, text: str) -> List[Aspect]:
        """Extract aspects using spaCy's dependency parsing.

        Args:
            text: Input text

        Returns:
            List of extracted aspects
        """
        if self._nlp is None:
            return []

        doc = self._nlp(text)
        aspects: List[Aspect] = []
        seen_texts: Set[str] = set()

        for token in doc:
            # Look for nouns that could be aspects
            if token.pos_ in ("NOUN", "PROPN"):
                # Get the noun phrase
                aspect_text = self._get_noun_phrase(token)

                # Skip if already seen
                if aspect_text.lower() in seen_texts:
                    continue

                # Detect sentiment from context
                sentiment = self._detect_sentiment_spacy(token, doc)

                # Calculate confidence based on POS patterns
                confidence = self._calculate_confidence_spacy(token)

                aspects.append(
                    Aspect(
                        text=aspect_text,
                        sentiment=sentiment,
                        confidence=confidence,
                        start_pos=token.idx,
                        end_pos=token.idx + len(aspect_text),
                    )
                )

                seen_texts.add(aspect_text.lower())

        return aspects

    def _get_noun_phrase(self, token) -> str:  # type: ignore
        """Extract noun phrase including modifiers.

        Args:
            token: spaCy token

        Returns:
            Complete noun phrase text
        """
        # Include adjective modifiers
        phrase_tokens = []

        # Add modifiers before the noun
        for child in token.lefts:
            if child.dep_ in ("amod", "compound", "nummod"):
                phrase_tokens.append(child.text)

        # Add the noun itself
        phrase_tokens.append(token.text)

        # Add modifiers after the noun
        for child in token.rights:
            if child.dep_ in ("compound", "nn"):
                phrase_tokens.append(child.text)

        return " ".join(phrase_tokens)

    def _detect_sentiment_spacy(self, token, doc) -> Optional[Sentiment]:  # type: ignore
        """Detect sentiment using dependency relations and lexicon.

        Args:
            token: Target token (aspect)
            doc: spaCy document

        Returns:
            Detected sentiment or None
        """
        # Check for sentiment words in the vicinity
        sentiment_words = []

        # Look at the token's head and children
        for related_token in list(token.children) + [token.head]:
            if related_token.text.lower() in self._sentiment_lexicon:
                sentiment_words.append(self._sentiment_lexicon[related_token.text.lower()])

        # Look for sentiment words in nearby adjectives
        for other_token in doc:
            if (
                other_token.pos_ == "ADJ"
                and abs(other_token.i - token.i) <= 3
                and other_token.text.lower() in self._sentiment_lexicon
            ):
                sentiment_words.append(self._sentiment_lexicon[other_token.text.lower()])

        # Determine overall sentiment
        if not sentiment_words:
            return None

        positive_count = sum(1 for s in sentiment_words if s == Sentiment.POSITIVE)
        negative_count = sum(1 for s in sentiment_words if s == Sentiment.NEGATIVE)

        if positive_count > 0 and negative_count > 0:
            return Sentiment.MIXED
        elif positive_count > 0:
            return Sentiment.POSITIVE
        elif negative_count > 0:
            return Sentiment.NEGATIVE
        else:
            return Sentiment.NEUTRAL

    def _calculate_confidence_spacy(self, token) -> float:  # type: ignore
        """Calculate confidence score based on linguistic features.

        Args:
            token: spaCy token

        Returns:
            Confidence score between 0 and 1
        """
        confidence = 0.7  # Base confidence

        # Higher confidence for nouns with modifiers
        if any(child.dep_ in ("amod", "compound") for child in token.children):
            confidence += 0.15

        # Higher confidence for subjects and objects
        if token.dep_ in ("nsubj", "dobj", "pobj"):
            confidence += 0.15

        return min(confidence, 1.0)

    def _extract_with_patterns(self, text: str) -> List[Aspect]:
        """Extract aspects using simple regex patterns (fallback).

        Args:
            text: Input text

        Returns:
            List of extracted aspects
        """
        aspects: List[Aspect] = []

        # Simple pattern: adjective + noun (e.g., "battery life", "camera quality")
        pattern = r"\b([a-zA-Z]+\s+(?:quality|life|performance|battery|camera|screen|design|price|sound|speed|size))\b"

        for match in re.finditer(pattern, text, re.IGNORECASE):
            aspect_text = match.group(1)

            # Detect sentiment from surrounding words
            context_start = max(0, match.start() - 30)
            context_end = min(len(text), match.end() + 30)
            context = text[context_start:context_end]

            sentiment = self._detect_sentiment_pattern(context)

            aspects.append(
                Aspect(
                    text=aspect_text,
                    sentiment=sentiment,
                    confidence=0.6,  # Lower confidence for pattern-based
                    start_pos=match.start(),
                    end_pos=match.end(),
                    context=context,
                )
            )

        return aspects

    def _detect_sentiment_pattern(self, context: str) -> Optional[Sentiment]:
        """Detect sentiment from context using lexicon.

        Args:
            context: Context text around aspect

        Returns:
            Detected sentiment or None
        """
        context_lower = context.lower()

        positive_count = sum(
            1
            for word in self._sentiment_lexicon
            if self._sentiment_lexicon[word] == Sentiment.POSITIVE and word in context_lower
        )

        negative_count = sum(
            1
            for word in self._sentiment_lexicon
            if self._sentiment_lexicon[word] == Sentiment.NEGATIVE and word in context_lower
        )

        if positive_count > 0 and negative_count > 0:
            return Sentiment.MIXED
        elif positive_count > 0:
            return Sentiment.POSITIVE
        elif negative_count > 0:
            return Sentiment.NEGATIVE
        else:
            return None

File: rust/aspect-extraction-core/src/extractors/rule_based.rs
//! Rule-based aspect extraction using linguistic patterns.
//!
//! This extractor uses regex patterns and simple sentiment analysis
//! to extract aspects from text without requiring ML models.

use crate::error::{Error, Result};
use crate::extractor::AspectExtractor;
use crate::types::{Aspect, Sentiment};
use crate::utils;
use regex::Regex;
use std::collections::HashMap;

/// Rule-based aspect extractor
pub struct RuleBasedExtractor {
    /// Minimum confidence threshold (0.0 to 1.0)
    min_confidence: f64,

    /// Regex patterns for aspect extraction
    patterns: Vec<Regex>,

    /// Sentiment lexicon (positive words)
    positive_words: HashMap<String, f64>,

    /// Sentiment lexicon (negative words)
    negative_words: HashMap<String, f64>,
}

impl RuleBasedExtractor {
    /// Create a new rule-based extractor with default configuration
    pub fn new() -> Self {
        Self::with_config(0.0)
    }

    /// Create a new rule-based extractor with custom configuration
    pub fn with_config(min_confidence: f64) -> Self {
        Self {
            min_confidence,
            patterns: Self::default_patterns(),
            positive_words: Self::default_positive_words(),
            negative_words: Self::default_negative_words(),
        }
    }

    /// Get default extraction patterns
    fn default_patterns() -> Vec<Regex> {
        vec![
            // Noun phrases: "the <aspect>"
            Regex::new(r"\b(the|a|an)\s+([a-z]+(?:\s+[a-z]+){0,2})\b").unwrap(),
            // Adjective + noun: "<adj> <aspect>"
            Regex::new(r"\b([a-z]+)\s+(quality|life|size|speed|performance|design|price)\b")
                .unwrap(),
            // Possessive: "its <aspect>", "the <aspect>"
            Regex::new(r"\b(its|their|my|your)\s+([a-z]+)\b").unwrap(),
        ]
    }

    /// Get default positive sentiment words
    fn default_positive_words() -> HashMap<String, f64> {
        let words = vec![
            ("good", 0.7),
            ("great", 0.9),
            ("excellent", 1.0),
            ("amazing", 1.0),
            ("fantastic", 1.0),
            ("awesome", 0.9),
            ("wonderful", 0.9),
            ("nice", 0.6),
            ("love", 1.0),
            ("perfect", 1.0),
            ("best", 1.0),
            ("outstanding", 1.0),
            ("superb", 0.9),
            ("brilliant", 0.9),
            ("positive", 0.7),
        ];

        words
            .into_iter()
            .map(|(w, s)| (w.to_string(), s))
            .collect()
    }

    /// Get default negative sentiment words
    fn default_negative_words() -> HashMap<String, f64> {
        let words = vec![
            ("bad", 0.7),
            ("poor", 0.8),
            ("terrible", 1.0),
            ("awful", 1.0),
            ("horrible", 1.0),
            ("disappointing", 0.8),
            ("worst", 1.0),
            ("hate", 1.0),
            ("useless", 0.9),
            ("pathetic", 0.9),
            ("negative", 0.7),
            ("mediocre", 0.6),
            ("inadequate", 0.7),
        ];

        words
            .into_iter()
            .map(|(w, s)| (w.to_string(), s))
            .collect()
    }

    /// Analyze sentiment for a given context
    fn analyze_sentiment(&self, context: &str) -> Option<Sentiment> {
        let tokens = utils::tokenize(context, true).ok()?;

        let mut pos_score = 0.0;
        let mut neg_score = 0.0;

        for token in &tokens {
            if let Some(score) = self.positive_words.get(token) {
                pos_score += score;
            }
            if let Some(score) = self.negative_words.get(token) {
                neg_score += score;
            }
        }

        if pos_score > neg_score && pos_score > 0.0 {
            Some(Sentiment::Positive)
        } else if neg_score > pos_score && neg_score > 0.0 {
            Some(Sentiment::Negative)
        } else if pos_score == 0.0 && neg_score == 0.0 {
            None
        } else {
            Some(Sentiment::Neutral)
        }
    }

    /// Extract context window around a match
    fn extract_context(&self, text: &str, start: usize, end: usize, window_size: usize) -> String {
        let context_start = start.saturating_sub(window_size);
        let context_end = (end + window_size).min(text.len());

        text[context_start..context_end].to_string()
    }

    /// Calculate confidence score based on pattern match and context
    fn calculate_confidence(&self, aspect_text: &str, context: &str) -> f64 {
        let mut confidence = 0.7; // Base confidence for pattern match

        // Boost if sentiment words are present
        let tokens = utils::tokenize(context, true).unwrap_or_default();
        let has_sentiment = tokens.iter().any(|t| {
            self.positive_words.contains_key(t) || self.negative_words.contains_key(t)
        });

        if has_sentiment {
            confidence += 0.2;
        }

        // Boost for longer aspects (more specific)
        let word_count = aspect_text.split_whitespace().count();
        if word_count >= 2 {
            confidence += 0.1;
        }

        confidence.min(1.0)
    }
}

impl Default for RuleBasedExtractor {
    fn default() -> Self {
        Self::new()
    }
}

impl AspectExtractor for RuleBasedExtractor {
    fn extract(&self, text: &str) -> Result<Vec<Aspect>> {
        let cleaned = utils::clean_text(text, true)?;

        let mut aspects = Vec::new();
        let mut seen = HashMap::new();

        for pattern in &self.patterns {
            for cap in pattern.captures_iter(&cleaned) {
                // Get the aspect text (usually the last capture group)
                let aspect_text = cap
                    .get(cap.len() - 1)
                    .map(|m| m.as_str())
                    .unwrap_or_default()
                    .trim()
                    .to_string();

                if aspect_text.is_empty() || aspect_text.len() < 2 {
                    continue;
                }

                // Skip if we've already seen this aspect
                if seen.contains_key(&aspect_text) {
                    continue;
                }

                // Get match position
                let match_obj = cap.get(0).unwrap();
                let start = match_obj.start();
                let end = match_obj.end();

                // Extract context
                let context = self.extract_context(&cleaned, start, end, 50);

                // Analyze sentiment
                let sentiment = self.analyze_sentiment(&context);

                // Calculate confidence
                let confidence = self.calculate_confidence(&aspect_text, &context);

                // Skip if below confidence threshold
                if confidence < self.min_confidence {
                    continue;
                }

                let aspect = Aspect::new(aspect_text.clone())
                    .sentiment(sentiment.unwrap_or(Sentiment::Neutral))
                    .position(start, end)
                    .context(context)
                    .confidence(confidence)?;

                seen.insert(aspect_text, ());
                aspects.push(aspect);
            }
        }

        Ok(aspects)
    }

    fn extract_batch(&self, texts: &[&str]) -> Result<Vec<Vec<Aspect>>> {
        // Parallel processing using rayon
        use rayon::prelude::*;

        texts
            .par_iter()
            .map(|text| self.extract(text))
            .collect()
    }

    fn name(&self) -> &'static str {
        "rule-based"
    }

    fn description(&self) -> &'static str {
        "Fast rule-based extraction using regex patterns and sentiment lexicon"
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_rule_based_extractor_new() {
        let extractor = RuleBasedExtractor::new();
        assert_eq!(extractor.min_confidence, 0.0);
    }

    #[test]
    fn test_extract_basic() {
        let extractor = RuleBasedExtractor::new();
        let text = "The camera quality is excellent but the battery life is poor";

        let aspects = extractor.extract(text).unwrap();

        assert!(!aspects.is_empty());
        // Should find "camera quality" and "battery life"
        let aspect_texts: Vec<&str> = aspects.iter().map(|a| a.text.as_str()).collect();
        assert!(aspect_texts.contains(&"camera quality") || aspect_texts.contains(&"quality"));
    }

    #[test]
    fn test_sentiment_analysis() {
        let extractor = RuleBasedExtractor::new();

        assert_eq!(
            extractor.analyze_sentiment("this is excellent"),
            Some(Sentiment::Positive)
        );
        assert_eq!(
            extractor.analyze_sentiment("this is terrible"),
            Some(Sentiment::Negative)
        );
        assert_eq!(extractor.analyze_sentiment("this is something"), None);
    }

    #[test]
    fn test_min_confidence_threshold() {
        let extractor = RuleBasedExtractor::with_config(0.95);
        let text = "The camera is nice";

        let aspects = extractor.extract(text).unwrap();

        // With high threshold, might filter out some aspects
        assert!(aspects.iter().all(|a| a.confidence >= 0.95));
    }

    #[test]
    fn test_extract_batch() {
        let extractor = RuleBasedExtractor::new();
        let texts = vec!["The camera is good", "The battery is bad"];

        let results = extractor.extract_batch(&texts).unwrap();

        assert_eq!(results.len(), 2);
    }

    #[test]
    fn test_extract_empty_text() {
        let extractor = RuleBasedExtractor::new();
        let result = extractor.extract("");

        assert!(result.is_err());
    }

    #[test]
    fn test_no_duplicate_aspects() {
        let extractor = RuleBasedExtractor::new();
        let text = "The battery is good and the battery is excellent";

        let aspects = extractor.extract(text).unwrap();

        // Should only extract "battery" once
        let aspect_texts: Vec<&str> = aspects.iter().map(|a| a.text.as_str()).collect();
        let battery_count = aspect_texts.iter().filter(|&&t| t == "battery").count();

        assert!(battery_count <= 1);
    }
}
